### PR TITLE
add beta_neg_binomial_lcdf

### PR DESCRIPTION
**Pull request by lingium**
## Summary

With this PR the lcdf of beta negative binomial distribution are added.
See issue #3115 

Fix a few typos in beta_neg_binomial_lccdf, pr #3114.

Expressions involved (almost the same as #3114 , just add the chain rule from log ccdf to log cdf):

$$
\begin{align}
& P(Y \leq y) = F(r,\alpha,\beta) = 1-C(r,\alpha,\beta) = 1-\frac{\Gamma (r+y +1) B(r+\alpha ,\beta +y +1) {}_3F_2(...)}{\Gamma (r) B(\alpha ,\beta ) \Gamma (y +2)}
\end{align}
$$

where $C(r,\alpha,\beta)$ is ccdf, we denote ${}_3F_2(\{1,r+y +1,\beta +y +1\}; \{y +2,r+\alpha +\beta +y +1\};1)$ as ${}_3F_2(...)$

Then the partial derivatives w.r.t log ccdf:

$$
\begin{align}
\frac{\partial \log C(r,\alpha,\beta)}{\partial r} &= \psi(r+y+1) + \psi(\alpha+r) - \psi(\alpha+\beta+r+y+1) - \psi(r) + \frac{\partial \log {}_3F_2(...)}{\partial r}, \\
\frac{\partial \log C(r,\alpha,\beta)}{\partial \alpha} &= \psi(\alpha+r) - \psi(\alpha+\beta+r+y+1) + \frac{\partial \log {}_3F_2(...)}{\partial \alpha} - \psi(\alpha) + \psi(\alpha+\beta), \\
\frac{\partial \log C(r,\alpha,\beta)}{\partial \beta} &= \psi(\beta+y+1) - \psi(\alpha+\beta+r+y+1) + \frac{\partial \log {}_3F_2(...)}{\partial \beta} - \psi(\beta) + \psi(\alpha+\beta).
\end{align}
$$

where

$$
\begin{align}
	\frac{\partial \log {}_3F_2(...)}{\partial r} &=  \frac{\partial {}_3F_2(...)}{\partial r} /  {}_3F_2(...) \\
	&= \left[ _3F_2(...)^{(\{0,1,0\},\{0,0\},0)}(...) + _3F_2(...)^{(\{0,0,0\},\{0,1\},0)}(...) \right]  / _3F_2(...), \\
\frac{\partial \log {}_3F_2(...)}{\partial \alpha} &=  {}_3F_2(...)^{(\{0,0,0\},\{0,1\},0)}(...)  / {}_3F_2(...), \\
\frac{\partial \log {}_3F_2(...)}{\partial \beta} &= \left[ _3F_2(...)^{(\{0,0,1\},\{0,0\},0)}(...) + _3F_2(...)^{(\{0,0,0\},\{0,1\},0)}(...) \right]  / _3F_2(...).
\end{align}
$$

By chain rule, 

$$
\begin{align}
\frac{\partial \log F(r,\alpha,\beta)}{\partial r} &= - \frac{1}{F(r,\alpha,\beta)} \cdot \frac{\partial C(r,\alpha,\beta)}{\partial r} &= - \frac{C(r,\alpha,\beta)}{1-C(r,\alpha,\beta)} \cdot \frac{\partial \log C(r,\alpha,\beta)}{\partial r}
\end{align}
$$

Therefore, the partial derivatives w.r.t log cdf:

$$
\begin{align}
\frac{\partial \log F(r,\alpha,\beta)}{\partial r} &= - \frac{C(r,\alpha,\beta)}{1-C(r,\alpha,\beta)} \cdot\frac{\partial \log C(r,\alpha,\beta)}{\partial r} , \\
\frac{\partial \log F(r,\alpha,\beta)}{\partial \alpha} &= - \frac{C(r,\alpha,\beta)}{1-C(r,\alpha,\beta)} \cdot\frac{\partial \log C(r,\alpha,\beta)}{\partial \alpha}, \\
\frac{\partial \log F(r,\alpha,\beta)}{\partial \beta} &= - \frac{C(r,\alpha,\beta)}{1-C(r,\alpha,\beta)} \cdot\frac{\partial \log C(r,\alpha,\beta)}{\partial \beta}.
\end{align}
$$

## Tests

Test is written follow the [guide](https://mc-stan.org/math/md_doxygen_2contributor__help__pages_2adding__new__distributions.html).

## Side Effects

No.

## Release notes

beta_neg_binomial_lcdf is available if merged.

## Checklist

- [x] Copyright holder: Zhi Ling

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Coding-Style-and-Idioms) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by WardBrian** [comment]
@lingium while it is very useful to keep each feature in a separate PR (thanks for doing this!), if you do have multiple things you want to add you can open several PRs at once, which might help with review lag-time. I think you said you also have an RNG for this distribution, and if so feel free to open a separate PR before this one is merged

**Comment by lingium** [comment]
Hi @WardBrian , got you, Just thought this modification is too minor so I didn't open another PR. I'll make sure to open separate PRs for different things next time. Yes, there is a RNG. I can open it in parallel.
BTW I'm encountering problems adding signatures to the compiler, still trying to understand the OCaml thing. Would you mind me to submit a PR anyway later and then you can tell me where the problem lies?

**Review by SteveBronder — DISMISSED** [review]
No notes!

File: test/prob/beta_neg_binomial/beta_neg_binomial_cdf_log_test.hpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  void valid_values(vector<vector<double>>& parameters,
+                    vector<double>& ccdf_log) {
+    vector<double> param(4);
+
+    param[0] = 10;   // n
+    param[1] = 5.5;  // r
+    param[2] = 2.5;  // alpha
+    param[3] = 0.5;  // beta
+    parameters.push_back(param);
+    ccdf_log.push_back(std::log(0.967906252841089));  // expected cdf_log
```
Comment (WardBrian) [inline test/prob/beta_neg_binomial/beta_neg_binomial_cdf_log_test.hpp:21]: Similar to the lccdf, can you find values for this that don't take as long? This looks like another 1day+ test

Comment (lingium) [reply test/prob/beta_neg_binomial/beta_neg_binomial_cdf_log_test.hpp:21]: Yep, tested on the local computer, the required numbers are the same as in `lccdf`.

**Review by WardBrian — APPROVED** [review]
Thanks, looks good!

**Pull request merged**